### PR TITLE
Added file metadata tests

### PR DESCRIPTION
**Pull request by Ramyapatil712**
## Purpose
Add the following checks to the selenium test automation suite:

1. First Test: Edit Metadata Tab
- Title field
- Description field
- Resource Type field
- Resource Language field

2. Second test: Download
- Download Metadata Button

## Summary of Changes

Added new test_metadata.py - 6 new tests added to verify title, description, resource type and resource language changes, cancel changes funtionality and download functionality
Updated project.py - added new FilesMetadataPage class , and ProjectMetadataPage class
Updated registries.py - updated RegistrationMetadataPage class
Updated osf_api.py — Added 2 new methods get_fake_file_guid and get_existing_file_data
Updated utils.py - Added 3 new methods verify_file_download, latest_download_file and get_guid_from_url

## Reviewer's Actions
`git fetch <remote> pull/265/head:feature/test_metadata`

Run this test using
`pytest tests/test_metadata.py -s -v`

## Testing Changes Moving Forward
N/A

## Ticket

https://openscience.atlassian.net/browse/ENG-4276
https://openscience.atlassian.net/browse/ENG-4277

### CODE REVIEW

File: tests/test_metadata.py [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+            session, project_guid=default_project_with_metadata.id
+        )
+        if funder_info is None:
+            WebDriverWait(driver, 5).until(
+                EC.element_to_be_clickable((By.XPATH, '//button[text()="Add funder"]'))
+            ).click()
+
+        project_metadata_page.funder_name.click()
+
+        project_metadata_page.funder_name_serach_input.send_keys(funder_name)
```
Comment (jh27539) [inline tests/test_metadata.py]: typo for 'search'

File: api/osf_api.py [hunk 2]
```diff
@@ -1050,3 +1050,136 @@ def get_project_node_analytics_data(session, node_id=None, timespan='week'):
...
+    """
+    if not session:
+        session = get_default_session()
+    url = '/v2/registrations/?embed=contributors'
+    page_data = session.get(url)['links']
+    total_registrations = page_data['meta']['total']
+    per_page = 10
+    total_pages = round(total_registrations / per_page)
+    for page in range(1, total_pages):
+        page_url = 'v2/registrations/?embed=contributors&page=%s' % page
```
Comment (jh27539) [inline api/osf_api.py]: let's try a search for title

Comment (jh27539) [reply api/osf_api.py]: Selenium Registration for Metadata tests

File: tests/test_metadata.py [hunk 3]
```diff
@@ -0,0 +1,716 @@
...
+        WebDriverWait(driver, 5).until(
+            EC.element_to_be_clickable(
+                (By.CSS_SELECTOR, '[data-test-edit-funding-metadata-button]')
+            )
+        ).click()
+
+        funder_info = osf_api.get_funder_data_project(
+            session, project_guid=default_project_with_metadata.id
+        )
+        if funder_info is None:
```
Comment (jh27539) [inline tests/test_metadata.py]: let's add the `else` statement

`        else:
            WebDriverWait(driver, 5).until(
                EC.element_to_be_clickable(
                    (By.XPATH, '//button[text()="Delete funder"]')
                )
            ).click()
            WebDriverWait(driver, 5).until(
                EC.element_to_be_clickable((By.XPATH, '//button[text()="Add funder"]'))
            ).click()`

File: tests/test_metadata.py [hunk 4]
```diff
@@ -0,0 +1,716 @@
...
+        osf_api.update_registration_metadata_with_custom_data(
+            registration_id=registration_guid
+        )
+        registration_metadata_page = RegistrationMetadataPage(
+            driver, guid=registration_guid
+        )
+        registration_metadata_page.goto()
+        return registration_metadata_page
+
+    @markers.smoke_test
```
Comment (jh27539) [inline tests/test_metadata.py]: Let's remove smoke tests for production

File: api/osf_api.py (hunk 2 above)
Comment (Ramyapatil712) [reply api/osf_api.py]: Done. Updated get_most_recent_registration_node_id_by_user to search with title (filtering with title in the url)

File: tests/test_metadata.py (hunk 3 above)
Comment (Ramyapatil712) [reply tests/test_metadata.py]: Done

File: tests/test_metadata.py (hunk 4 above)
Comment (Ramyapatil712) [reply tests/test_metadata.py]: Done.Added markers on the parent

File: api/osf_api.py [hunk 5]
```diff
@@ -1104,45 +1104,20 @@ def delete_project_contributor(session, node_id, user_name):
...
-                            return registration['id']
-                            break
-                    else:
-                        continue
-                    break
+    session = client.Session(
+        api_base_url=settings.API_DOMAIN,
+        auth=(settings.REGISTRATIONS_USER, settings.REGISTRATIONS_USER_PASSWORD),
+    )
+    url = '/v2/registrations/?filter[title][icontains]=selenium'
```
Comment (jh27539) [inline api/osf_api.py]: are there multiple registrations with 'selenium' as the title?

Comment (jh27539) [reply api/osf_api.py:1115]: https://www.urlencoder.org/

Comment (jh27539) [reply api/osf_api.py]: In the console, `encodeURIComponent('selenium test registration')`

File: tests/test_metadata.py [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+        registration_metadata_page.award_info_URI.click()
+        registration_metadata_page.award_info_URI.send_keys(award_info_uri)
+        registration_metadata_page.award_number.click()
+        registration_metadata_page.award_number.send_keys(award_number)
+        registration_metadata_page.add_funder_button.click()
+
+        registration_metadata_page.scroll_into_view(
+            registration_metadata_page.delete_funder_button.element
+        )
+        registration_metadata_page.delete_funder_button.click()
```
Comment (jh27539) [inline tests/test_metadata.py:637]: Selenium expected condition `element_to_be_clickable` will be helpful here

File: tests/test_metadata.py [hunk 7]
```diff
@@ -0,0 +1,708 @@
...
+        project_metadata_page.award_title.send_keys(award_title)
+        project_metadata_page.award_info_URI.click()
+        project_metadata_page.award_info_URI.send_keys(award_uri)
+        project_metadata_page.award_number.click()
+        project_metadata_page.award_number.send_keys(award_number)
+        project_metadata_page.add_funder_button.click()
+        project_metadata_page.scroll_into_view(
+            project_metadata_page.delete_funder_button.element
+        )
+        project_metadata_page.delete_funder_button.click()
```
Comment (jh27539) [inline tests/test_metadata.py:494]: use a `element_to_be_clickable`

File: api/osf_api.py [hunk 8]
```diff
@@ -1173,3 +1155,20 @@ def get_funder_data_registration(registration_guid):
...
+        api_base_url=settings.API_DOMAIN,
+        auth=(settings.REGISTRATIONS_USER, settings.REGISTRATIONS_USER_PASSWORD),
+    )
+
+    url = '/v2/registrations/?filter[title]=' + encoded_registration_title
+    data = session.get(url)['data']
+    if data:
+        return data[0]['id']
+
+    return None
```
Comment (cslzchen) [inline api/osf_api.py:1174]: No a blocker, some coding style suggestion

```
return data[0]['id'] if data else None
```

File: utils.py [hunk 9]
```diff
@@ -127,3 +130,76 @@ def close_current_tab(driver, main_window):
...
+
+def find_row_by_name(files_page, file_name):
+    all_files = files_page.file_rows
+    for file_row in all_files:
+        if file_name in file_row.text:
+            return file_row
+    return
+
+
+def verify_file_download(driver, file_name):
```
Comment (cslzchen) [inline utils.py:143]: 👍 love this feature 😄

File: api/osf_api.py [hunk 10]
```diff
@@ -1050,3 +1050,125 @@ def get_project_node_analytics_data(session, node_id=None, timespan='week'):
...
+            'attributes': {
+                'language': 'eng',
+                'resource_type_general': 'Collection',
+            },
+        }
+    }
+    session.put(url=url, raw_body=json.dumps(raw_payload))
+
+
+def delete_project_contributor(session, node_id, user_name):
```
Comment (cslzchen) [inline api/osf_api.py:1092]: Just a comment: if contributor is the only admin, this api request will fail. I don't think it affects what we are doing here in the tests.

File: settings.py [hunk 11]
```diff
@@ -54,6 +54,7 @@
 
 NEW_USER_EMAIL = env('NEW_USER_EMAIL')
 
+
```
Comment (cslzchen) [inline settings.py:57]: Remove this extra empty line

File: tests/test_metadata.py [hunk 12]
```diff
@@ -0,0 +1,709 @@
...
+    data = response.json()
+    for funder in data['included']:
+        if funder['type'] == 'index-card':
+            if (
+                funder['attributes']['resourceMetadata']['name'][0]['@value']
+                == funder_name
+            ):
+                award_title = funder['attributes']['resourceMetadata']['resourceType'][
+                    0
+                ]['@id']
```
Comment (cslzchen) [inline tests/test_metadata.py]: Out of curiosity, is this style automatically done by `black` and/or `isort`? It looks way off for me though.

File: tests/test_metadata.py [hunk 13]
```diff
@@ -0,0 +1,709 @@
...
+
+        # Select 'Book' from the resource type listbox
+        file_metadata_page.resource_type.click()
+        WebDriverWait(driver, 5).until(
+            EC.visibility_of_element_located(
+                (
+                    By.CSS_SELECTOR,
+                    '#ember-basic-dropdown-wormhole > div > ul > li>span',
+                )
+            )
```
Comment (cslzchen) [inline tests/test_metadata.py]: The above style might be an isort/black thing, but if not, I suggest the following:

```python
EC.visibility_of_element_located((
        By.CSS_SELECTOR,
        '#ember-basic-dropdown-wormhole > div > ul > li>span',
))
```

Comment (cslzchen) [reply tests/test_metadata.py]: yeah, I see, this happens in lots of places, not really worth the change, just treat my style suggestions as comments :)

**Review by cslzchen — APPROVED** [review]
LGTM after you respond to Josh's two comments. Mine are purely style suggestions. 🎆

**Review by cslzchen — APPROVED** [review]

**Pull request merged**